streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -505,4 +550,35 @@ private StateStoreMetadata findStore(final TopicPartition changelogPartition) {
 
         return found.isEmpty() ? null : found.get(0);
     }
+
+    void prepareForRecycle() {
+        log.debug("Preparing to recycle state for {} task {}.", taskType, taskId);
+
+        if (recyclingState) {
+            throw new IllegalStateException("Attempted to re-recycle state without completing first recycle");
+        }
+        recyclingState = true;
+    }
+
+    void recycleState() {
+        log.debug("Completed recycling state for formerly {} task {}.", taskType, taskId);
+
+        if (!recyclingState) {
+            throw new IllegalStateException("Attempted to complete recycle but state is not currently being recycled");
+        }
+        recyclingState = false;
+
+        if (taskType == TaskType.ACTIVE) {
+            taskType = TaskType.STANDBY;
+        } else {
+            taskType = TaskType.ACTIVE;
+        }
```

Comment:
It seems a bit "bold" here to assume that we want to flip the active/standby state. The TaskManager objectively knows what type it wants the task to become, so it seems it should just inform us of the desired task type in `prepareForRecycle`.

Or, better yet, just null it out and we can set it during `createStandbyTaskFromActive` and `createActiveTaskFromStandby`. This kind of thing might also be nice, since we could have a more explicit lifecycle for this object:
* I'm ready (created, valid, good to go)
* I'm getting recycled
* I'm closed

Then, we can ensure that these "createXFromY" methods cleanly take the state manager from "closed" to "ready".
I'm not saying to add _another_ state enum, but having a clearly defined lifecycle will help us later in maintenance.

Reply:
That makes sense to me, I'll try it out

Reply:
Alright I reworked things slightly here. With nulling out the task type during close/recycle we can technically get away with not even having the `recyclingState` flag at all. But I thought it might be a good idea to keep the two distinct to avoid future messes

Reply:
Actually turns out nulling it during close causes some tests to fail 😕 Probably the changelog reader.  I'll just revert that part for now and look into it next week